common/src/main/java/com/wynntils/features/user/redirects/BlacksmithRedirectFeature.java

```diff
@@ -0,0 +1,102 @@
+        // scrap we get.
+        String paymentString = "";
+        Matcher messageMatcher = BLACKSMITH_MESSAGE_PATTERN.matcher(event.getCodedMessage());
+        if (!messageMatcher.matches()) return;
+        event.setCanceled(true);
+
+        // Retrieve the color code of the item, and then match it to the item tier.
+        Matcher itemMatcher = ITEM_PATTERN.matcher(messageMatcher.group(2)); // Second group contains all of the items.
+        while (itemMatcher.find()) {
+            ChatFormatting itemColor = ChatFormatting.getByCode(itemMatcher
+                    .group(1)
+                    .charAt(0)); // find the color code, then get the ChatFormatting of the item to match it to a tier.
+            ItemTier tierToIncrease = ItemTier.fromChatFormatting(itemColor);
+            if (tierToIncrease == null) continue;
+            totalItems.put(tierToIncrease, totalItems.getOrDefault(tierToIncrease, 0) + 1);
+            totalItemInteger++;
+        }
+
+        // The final part of the message.
+        {
+            // Let's tally up the total number of items sold.
+            StringBuilder messageCounts = new StringBuilder();
+            for (ItemTier tier : ItemTier.values()) {
+                messageCounts.append('/' + tier.getChatFormatting().toString() + totalItems.getOrDefault(tier, 0));
+                messageCounts.append(ChatFormatting.LIGHT_PURPLE);
+            }
+
+            String pluralizer = totalItemInteger == 1 ? "" : "s";
+
+            messageCounts.append(") item" + pluralizer + " for ");
+
+            messageCounts.setCharAt(0, '(');
+
+            // How many emeralds/scrap we got from the blacksmith.
+            paymentString = messageMatcher.group(3);
+
+            String messageCountsString = messageCounts.toString();
+
+            // Now, we create the full message.
+            StringBuilder sendableMessage = new StringBuilder();
+            // Normal sale of items for emeralds.
+            if (messageMatcher.group(1).equals("sold me")) {
+                sendableMessage.append(ChatFormatting.LIGHT_PURPLE + "Sold " + totalItemInteger + " ");
+                sendableMessage.append(messageCountsString);
+                sendableMessage.append(ChatFormatting.GREEN + paymentString + EmeraldSymbols.EMERALDS
+                        + ChatFormatting.LIGHT_PURPLE + ".");
+            }
+            // Scrapping items for scrap
+            else {
+                sendableMessage.append(ChatFormatting.LIGHT_PURPLE + "Scrapped " + totalItemInteger);
+                sendableMessage.append(" item" + pluralizer + " for ");
+                sendableMessage.append(
+                        ChatFormatting.YELLOW + paymentString + " scrap" + ChatFormatting.LIGHT_PURPLE + ".");
+            }
+
+            String sendableMessageString = sendableMessage.toString();
+
+            // Finally, we send the message.
+            NotificationManager.queueMessage(sendableMessageString);
```

Comment:
I am not happy about this whole string generation part, it is a bit of a mess. 

Let's refactor this:
1, Create a StringBuilder to create the sold items (`1/0/0/0/0` part)
2, If we sold items -> "Sold %d (%s) item(s) for %dE",
    otherwise -> "Scrapped %d (%s) item(s) for %dscrap",